### PR TITLE
[WIP] new offset management to support retrying messages

### PR DESCRIPTION
**Pull request by achille-roussel**
This PR will track the work I'm doing to add support for retries on the Reader, it is still a work in progress.

The first step I'm taking is to add the `OffsetStore` abstraction, which represents the data store where the offsets will be tracked. I added an in-memory implementation of the store that acted as a reference to run the test against, and another implementation which uses a Kafka partition to store the offsets (`OffsetLog`).

The offset data structure is a simple pair of a Kafka offset value, and a timestamp. This should be enough to track the messages that need to be rescheduled and the time at which it needs to happen. I'm picturing that the offset topic would be compacted, with a number of partitions and a retention equal to the number of partitions in the data topic (with a 1:1 mapping of data to offset partitions).

The next steps are for me to add some kind of priority queue that will be used to track the retry times of the offsets, then the Reader will combine the offset store and the queue to construct the retry engine.

The data stored in the offset topic is really small, basically each message is a pair of two 64 bits integers (one for the key - which is the offset of the message to track in the data topic, and one of the value - which is the timestamp for the retry time). This means that we should be able to replay millions of offsets very quickly when we need to recreate the in-memory data structures, which would happen when a consumer gains ownership of a partition.

Please let me know if you see any major flaw in the design, or if you have any comment on the code.

### CODE REVIEW

**Comment by achille-roussel** [comment]
Alright this is coming close to be feature complete. I'm not super thrilled with the increased complexity in the reader and there are probably implementation details that could be improved (I tried to left notes in the code every time I had a doubt).

**Couple of changes I made along the way**

I made the whole implementation completely hidden from the program, all it gets to control are three parameters being `RequeueTopic`, `RequeuePeriod`, and `RequeueFetchMaxUnusedRatio` (see the code for documentation). I think it's safer to keep the change as local to kafka-go as possible to avoid introducing public APIs that we'd have to take back if they don't work well later on.

I made a couple of internal changes to existing data structures that were related to the changes I was making, tests are still passing so far but it does mean the code would deserve a careful review.

Overall I think the reader code would deserve a bit of cleanup and a freshen up design to make it easier to evolve (with or without this PR), if we have time some days it would be good to clean this up.

**Next steps here**

- Add unit tests to verify the basic behavior of this code.
- Add counters to the reader stats to track important events of the requeue mechanism.
- Build a test program to run this against our kafka deployments.

**Some things I'm unsure about**

The current implementation has no limit on how many requeues may be pending, which is probably a recipe for massive failure 😛I think there are a couple of approaches we can take here, I'd like to get your take on them:
- Limit how many requeues there can be, and stop consuming the live stream when we reach it. This may be tricky to do well tho because it may prevent consumers recovering from lag if the count of requeues does not go down quickly enough due to backoff policies of the application for example.
- Prioritize requeues over live traffic, this could help put back pressure on the live stream while allowing requeues to flush when they are scheduled, but it wouldn't prevent live data from going through when the reader has no requeues to scheduled for a period.
- Drop requeues when they reach the configured limit, so they act as a best-effort solution and the only garantee is the reader will produce each message at least once.
- ???

**Comment by achille-roussel** [comment]
Something else that I wanted feedback on, the files structure. I'm tempted to merge a couple of files to avoid the proliferation that's happening in this PR, but I'd like to get feedback on how the code reads currently and whether that would improve readability or not.

Let me know!

**Comment by stevevls** [comment]
I did an initial pass, but I need to digest and do further review before leaving feedback...there's a lot going on!

File: offsettimeline.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+func (q *offsetTimelineQueue) Less(i, j int) bool {
+	return q.offsets[i].time < q.offsets[j].time
+}
+
+func (q *offsetTimelineQueue) Swap(i, j int) {
+	q.offsets[i], q.offsets[j] = q.offsets[j], q.offsets[i]
+}
+
+func (q *offsetTimelineQueue) Push(x interface{}) {
```
Comment (rjenkins) [inline offsettimeline.go]: Is it ok that Len/Less/Swap and Push aren't protected by mutex's?

**Comment by achille-roussel** [comment]
@Pryz yes definitely, this is kind of already implemented (although probably needs a test or two to flush out edge cases). I'm more concerned about the amount of requeues  within the offset retention time (let's say a consumer fails 100% of events for a couple of minutes, this could blow up memory quickly).

File: offset.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	return off.toKey(), off.toValue()
+}
+
+type offsetKey int64
+
+func (k *offsetKey) readFrom(b []byte) {
+	*k = offsetKey(binary.BigEndian.Uint64(b))
+}
+
+func (k offsetKey) writeTo(b []byte) {
```
[the remark this answers is not in the document]
Comment (achille-roussel) [reply offset.go]: It doesn't save anything, I can change it if you think it makes more sense.

File: offset.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	_ = b[23] // to help with bound check elimination
+	*v = offsetValue{
+		attempt: int64(binary.BigEndian.Uint64(b[:8])),
+		size:    int64(binary.BigEndian.Uint64(b[8:16])),
+		time:    int64(binary.BigEndian.Uint64(b[16:24])),
+	}
+}
+
+func (v offsetValue) writeTo(b []byte) {
+	_ = b[23] // to help with bound check elimination
```
[the remark this answers is not in the document]
Comment (achille-roussel) [reply offset.go]: Indeed.

File: offsetlog.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+// offsetLog is an implementation of the offsetStore interface which uses a
+// Kafka partition as backing store for the offsets.
+//
+// Ideally the partition belongs to a compacted topic, but this is not required
+// and the log will work just fine with a regular topic (replays may take a bit
+// longer if there are large numbers of offsets recorded in the log).
+type offsetLog struct {
+	// Addresses of the Kafka brokers tha the log uses to lookup the leader of
+	// the topic partition that it uses.
+	brokers []string
```
[the remark this answers is not in the document]
Comment (achille-roussel) [reply offsetlog.go]: I'm sorry I'm not understanding what you mean here, do you mind giving a bit more details?

File: offsetlog.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+	// Name of the Kafka topic that the offset log uses to keep track of the
+	// offsets.
+	topic string
+
+	// Partition tha the offset log uses in the Kafka topic.
+	partition int
+
+	// Size limit of the fetch requests to the offset log, default to 8 MB.
+	maxBytes int
```
[the remark this answers is not in the document]
Comment (achille-roussel) [reply offsetlog.go]: No reason, `int` is used across the package for partition and byte counts, so I just kept it consistent.

File: offsetlog.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+const (
+	defaultOffsetLogReadTimeout  = 3 * time.Second
+	defaultOffsetLogWriteTimeout = 3 * time.Second
+	defaultOffsetLogMaxBytes     = 8 * 1024 * 1024
+)
+
+// readOffsets satisfies the offsetStore interface, it returns a list of offsets
+// in the partition read by the log.
+func (log *offsetLog) readOffsets() ([]offset, error) {
+	c, err := log.dialLeader()
```
[the remark this answers is not in the document]
Comment (achille-roussel) [reply offsetlog.go]: We don't, we could pool the connections.

For the read operation it doesn't matter much because it only happens once when the consumer starts reading from a partition in order to sync the requeued offsets.

For the writeOffsets and deleteOffsets operations we'll probably want to reuse connections, I just wanted to get something simple working first.

File: offsettimeline.go (hunk 1 above)
Comment (achille-roussel) [reply offsettimeline.go]: Yes, synchronization is done on the `offsetTimeline` value, `offsetTimelineQueue` is an internal type which is not manipulated concurrently.

File: offsetlog.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rjenkins) [reply offsetlog.go]: nm I misread and thought this was on each retry message.

File: offsetlog.go [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+func (log *offsetLog) dialLeader() (*Conn, error) {
+	if len(log.brokers) == 0 {
+		return nil, errors.New("no brokers configured on the offset log")
+	}
+	dialer := log.dialer
+	if dialer == nil {
+		dialer = DefaultDialer
+	}
+	broker := log.brokers[rand.Intn(len(log.brokers))]
+	return dialer.DialLeader(context.Background(), "tcp", broker, log.topic, log.partition)
```
Comment (rjenkins) [inline offsetlog.go]: Ideally would we retry on a different broker, if the random broker you fetched from the bootstrap did not connect, until all were exhausted?

**Comment by stevevls** [comment]
Still making my way through this PR...I've gone through it twice so far and am starting to see how it all hangs together.  One thing that I think would help greatly in the `Reader` code would be to use a different term than `unqueue` for the success case.  So many things have the name "queue" in them that I find it a bit hard to grok.  Something that indicates the successful nature of those operations (`markSucceeded`?) would clear that up I think!

**Comment by achille-roussel** [comment]
The work we're doing on Centrifuge is going to be a superset, and more robust solution to this problem, so I'll close this PR.

Thanks for all the feedback!
